### PR TITLE
TSIG Verify/Generate using TsigProvider

### PR DESCRIPTION
**Pull request by jelu**
By exposing `TsigVerifyProvider` and `TsigGenerateProvider` functions I am able to fully implement TSIG support in [DNS-OARC/golang-dns-server-doq](https://github.com/DNS-OARC/golang-dns-server-doq) similar to how it's done in `dns.Server` and without needed to copy almost all of `tsig.go`.

I see no harm in exposing these functions and maybe others will find them useful also, hope it can be merged.

### CODE REVIEW

**Comment by miekg** [comment]
I can't find a good reason to say 'no', so likely to be merged.

as the function docs are now public, some comments there.

**Comment by miekg** [comment]
I can't find a good reason to say 'no', so likely to be merged.

as the function docs are now public, some comments there.

File: tsig.go [hunk 1]
```diff
@@ -166,10 +166,12 @@ type timerWireFmt struct {
 // timersOnly is false.
 // If something goes wrong an error is returned, otherwise it is nil.
 func TsigGenerate(m *Msg, secret, requestMAC string, timersOnly bool) ([]byte, string, error) {
-	return tsigGenerateProvider(m, tsigHMACProvider(secret), requestMAC, timersOnly)
+	return TsigGenerateProvider(m, tsigHMACProvider(secret), requestMAC, timersOnly)
 }
 
-func tsigGenerateProvider(m *Msg, provider TsigProvider, requestMAC string, timersOnly bool) ([]byte, string, error) {
+// TsigGenerate fills out the TSIG record attached to the message using
+// a TsigProvider, for more details and return see TsigGenerate.
```
Comment (miekg) [inline tsig.go:173]: also unfinished docs

File: tsig.go [hunk 2]
```diff
@@ -166,10 +166,12 @@ type timerWireFmt struct {
 // timersOnly is false.
 // If something goes wrong an error is returned, otherwise it is nil.
 func TsigGenerate(m *Msg, secret, requestMAC string, timersOnly bool) ([]byte, string, error) {
```
Comment (miekg) [inline tsig.go:168]: on the docs - it's weirdly formatted, also Tsig RR -> TSIG RR..

"... called for the first time requestMAC should be set to the empty string and timersOnly should be false.":

would also be nice to tell what subsequent calls should look like,.

The "if something goes wrong an error is returned..." can be removed IMO

File: tsig.go [hunk 3]
```diff
@@ -223,7 +225,9 @@ func TsigVerify(msg []byte, secret, requestMAC string, timersOnly bool) error {
 	return tsigVerify(msg, tsigHMACProvider(secret), requestMAC, timersOnly, uint64(time.Now().Unix()))
 }
 
-func tsigVerifyProvider(msg []byte, provider TsigProvider, requestMAC string, timersOnly bool) error {
+// TsigVerify verifies the TSIG on a message using a TsigProvider, for
+// more details and return see TsigVerify.
```
[the remark this answers is not in the document]
Comment (jelu) [reply tsig.go:229]: It says `see TsigVerify`(?)

File: tsig.go (hunk 1 above)
Comment (jelu) [reply tsig.go:173]: same here

File: tsig.go (hunk 2 above)
Comment (jelu) [reply tsig.go:168]: I have made no change in the documentation for that function.

**Comment by jelu** [comment]
The return of the functions are the same, do you really want duplicated documentation?

**Comment by jelu** [comment]
Could anyone please let me know what kind of timeline I should expect for getting this merged and into a release?

Just want to know so I can plan the release of my library that depends on this PR.

**Comment by jelu** [comment]
> The return of the functions are the same, do you really want duplicated documentation?

@miekg Do you really want duplicated documentation?

**Comment by miekg** [comment]
this lgtm although `WithProvider` would be more Go-like and we already have some other 'With...' functions. 

I'll close this and send in a new PR with that name in it *and* also adjust any docs

**Comment by jelu** [comment]
OK.

It's totally okay for me if you push onto this branch also, it should be
open for you.

**Comment by miekg** [comment]
https://github.com/miekg/dns/pull/1382

please double check for anything stupid

**Comment by miekg** [comment]
Ah, must have misread that.

Ok..I'll push through a separate PR to clean that wording a bit.

**Pull request closed without merging**